experimental/oak_async/tests/tests.rs

```diff
@@ -0,0 +1,287 @@
+            };
+            add_ready_data(handle.handle(), &DummyData::new(data));
+            handle.set_status(ChannelReadStatus::ReadReady);
+        }
+        OakStatus::Ok
+    });
+
+    let (a, b): (Result<DummyData, OakError>, Result<DummyData, OakError>) = block_on(
+        futures::future::join(do_channel_read(HANDLE_0), do_channel_read(HANDLE_1)),
+    )
+    .expect("block_on failed");
+
+    assert_eq!(
+        a.expect("Read from channel HANDLE_0 failed"),
+        DummyData::new("Hello")
+    );
+
+    assert_eq!(
+        b.expect("Read from channel HANDLE_1 failed"),
+        DummyData::new("world!")
+    );
+}
+
+#[test]
+fn block_on_drop_channel_read_immediately() {
+    init();
+    block_on(async {
+        // The future is immediately dropped, so never polled
+        let _ = do_channel_read::<DummyData>(1);
+    })
+    .expect("block_on failed");
+}
+
+#[test]
+fn block_on_drop_channel_read_after_wait() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        std::thread::sleep(std::time::Duration::from_secs(1));
+        for handle in handles.iter_mut() {
+            match handle.handle() {
+                HANDLE_0 => {
+                    handle.set_status(ChannelReadStatus::NotReady);
+                }
+                HANDLE_1 => {
+                    add_ready_data(HANDLE_1, &DummyData::new("hello"));
+                    handle.set_status(ChannelReadStatus::ReadReady);
+                }
+                c => panic!("Unexpected waiting channel {}", c),
+            };
+        }
+        OakStatus::Ok
+    });
+
+    let result: Result<DummyData, OakError> = block_on(async {
+        futures::future::select(do_channel_read(HANDLE_0), do_channel_read(HANDLE_1))
+            .await
+            .factor_first()
+            // Note: It is important that we extract the result inside the async block. Doing this
+            // outside the async scope will block indefinitely, because it will try and move the
+            // unresolved ChannelRead on HANDLE_0 outside of the async block, but for block_on to
+            // terminate that ChannelRead must be `Drop`ed first (to remove it from the waiting
+            // set).
+            .0
+    })
+    .expect("block_on failed");
+
+    assert_eq!(
+        result.expect("Channel read failed"),
+        DummyData::new("hello")
+    );
+}
+
+#[test]
+fn block_on_channel_read_propagate_immediate_error() {
+    init();
+    set_error(666, OakStatus::ErrBadHandle);
+
+    let result: Result<DummyData, _> = block_on(do_channel_read(666)).expect("block_on failed");
+
+    match result {
+        Err(OakError::OakStatus(OakStatus::ErrBadHandle)) => { /* The expected error */ }
+        e => panic!("Unexpected error: {:?}", e),
+    }
+}
+
+#[test]
+fn block_on_channel_read_propagate_error_after_wait() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        assert_eq!(handles.len(), 1);
+        assert_eq!(handles[0].handle(), 666);
```

Comment:
Why not use HANDLE_0 or HANDLE_1 too? Or make 666 HANDLE_2, I guess?

Reply:
Whoops, missed this one. Will change to `HANDLE_0`